app/store/middleware.js

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import DeviceInfo from 'react-native-device-info';
+
+import {ViewTypes} from 'app/constants';
+import Config from 'assets/config';
+
+export function messageRetention() {
+    return (next) => (action) => {
+        if (action.type === 'persist/REHYDRATE') {
+            const {app} = action.payload;
+            const {entities} = action.payload;
+
+            if (!entities) {
+                return next(action);
+            }
+
+            // When a new version of the app has been detected
+            if (!app || !app.version || app.version !== DeviceInfo.getVersion() || app.build !== DeviceInfo.getBuildNumber()) {
+                return next(resetStateForNewVersion(action));
+            }
+
+            // Keep only the last 60 messages for the last 5 viewed channels in each team
+            // and apply data retention on those posts if applies
+            return next(cleanupState(action));
+        } else if (action.type === ViewTypes.DATA_CLEANUP) {
+            const nextAction = cleanupState(action, true);
+            return next(nextAction);
+        }
+
+        return next(action);
+    };
+}
+
+function resetStateForNewVersion(action) {
+    const {payload} = action;
+    const lastChannelForTeam = getLastChannelForTeam(payload);
+    const currentUserId = payload.entities.users.currentUserId;
+
+    const nextState = {
```

Comment:
Would it make sense to do this so that it'll keeps fields by default? Right now, we'll have to remember to add any fields we want to keep here

Reply:
well that's the whole point

Reply:
Fine by me then. It might be better to do that when persisting to the store in the first place so we can stop needing to blacklist stuff